### ISSUE TITLE
Allow `ETags` for HTML files

### ISSUE THREAD

**Issue by ThomasR** [h5bp/server-configs-apache#40]
Current situation in `.htaccess`:

```
<IfModule mod_headers.c>
    Header unset ETag
</IfModule>
FileETag None
ExpiresByType text/html "access plus 0 seconds"
```

This forces the client to reload HTML files regardless of whether or not they have changed.

I propose this:

```
<FilesMatch "(?<!\.html)$">
    <ifModule mod_headers.c>
        Header unset ETag
    </IfModule>
    FileETag None
</FilesMatch>
```

This will make apache generate ETags for HTML files.

Now, in the rare(?) case of server clustering, the client may still be forced to reload as mentioned in https://developer.yahoo.com/performance/rules.html#etags. Without server clustering the client benefits from ETags, and receives a 304 response.

Bottom line: most users will take advantage of this configuration, and the situation is unchanged for the rest.

**Comment by hatzopoulos** [comment]
If one wanted the text/html files cached wouldn't they modify `ExpiresByType text/html "access plus 0 seconds"` to their personal setting and thus not require etags?

I sometimes masquerade the .php files as .html files so I personally would not want ETags enabled or caching of those files.

**Comment by ThomasR** [comment]
@hatzopoulus
I agree that ExpiresByType should be set too, but that does not eliminate the need for ETags. See e.g. http://stackoverflow.com/a/500103 for a nice explanation.

**Comment by ThomasR** [comment]
I forgot to mention that the Expires header should be set to some far future value, not 0, when ETags are used.

**Comment by Malvoz** [comment]
In addition, [Mark Nottingham recommends](https://www.mnot.net/blog/2014/06/01/chrome_and_stale-while-revalidate) setting [`stale-while-revalidate`](https://tools.ietf.org/html/rfc5861#section-3) for (e.g.) HTML pages with a long `max-age` to "_hide the refresh latency on subsequent visits_".

**Comment by Malvoz** [comment]
My example approach:

```
<FilesMatch "(?<!\.html)$">

  <IfModule mod_headers.c>
    Header set Cache-Control "max-age=2592000, stale-while-revalidate=864000"
    Header unset ETag
  </IfModule>

  FileETag None

  <IfModule mod_expires.c>
    ExpiresActive Off
  </IfModule>

</FilesMatch>
```

**Comment by Malvoz** [comment]
There's not much movement on this issue, maybe unknowingly (hasn't been stated here) for good reasons - currently h5bp is compressing HTML files (which is nice), but GZIP and ETags seems to have issues when used in conjuction with one another. 

Short explainer: http://scarff.id.au/blog/2009/apache-304-and-mod_deflate-revisited/ 


ETag over compression should be worth on subsequent visits, but not compressing on first visits may lead to a negative user experience.
Unless there is a solution to using both GZIP and Etag, this issue should probably close.

**Comment by LeoColomb** [comment]
@Malvoz Thanks a lot for all these additions, good to know.
We will keep the issue opened as ETags are definitely headers to consider. Maybe not yet if there are some issues, but in the future.

Thanks again! 👍

**Comment by Malvoz** [comment]
Also from http://scarff.id.au/blog/2009/apache-304-and-mod_deflate-revisited/:
> As [Roy Fielding pointed out](https://issues.apache.org/bugzilla/show_bug.cgi?id=39727#c31) the correct way to deal with this issue is to stop abusing Content-Encoding for performance-compression and start using **Transfer-Encoding**; pity browsers and HTTP servers haven’t got there yet.

That was written in 2009, `Transfer-Encoding` seems to have full support (see browser compatability table here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding).

`Transfer-Encoding` + `ETag` = ❤️ ?

Anyone want to help investigate this?

**Comment by LeoColomb** [comment]
Config is ready for ETags via 23fc8b3.
Thing is [`DeflateAlterETag` directive](https://httpd.apache.org/docs/trunk/mod/mod_deflate.html#deflatealteretag), required for a working config, is not part of the stable releases of httpd so far, but planned for v2.5.

On hold.
